### DISCUSSION TITLE
What is the most stable Adobe Air version right now?

### DISCUSSION THREAD

**Discussion by premiumsB** [airsdk/Adobe-Runtime-Support#3040] [General]
Hello

I'm a bit tired to try version of Adobe Air and I was wondering which version is the most stable right now for IOS and Android?

Right now I use 50.2.2.6. Any earlier version reliable you tested so far?

Thank you for your answer.
Robert

**Comment by lwmirkk** [comment]
In Android I use this version as the most stable: 50.2.4.1 - not tested the 50.2.4.3 yet
In Windows, as my main usage is video player, I have to yet to use the 32.0.0.89 ;(
Tested every update in Windows so far and yet is not the best to play every MP4 video I tested. I will wait the new video pipeline, so this could be better.

I could not say anything in iOS.

**Comment by premiumsB** [comment]
@lwmirkk Thank you for your answer. Anybody for IOS? I hear that if you build your application with windows you cannot use several version of AIR because it may cause error with IPHONE 14 and IPHONE 15. 

Which versions do you guys use to build your IPA with windows?

**Reply by itlancer** [reply]
@premiumsB, I use 50.2.4.1 for all platforms (for iOS via Windows build too). 50.2.4.2-50.2.4.3 for me has critical issue for Android: https://github.com/airsdk/Adobe-Runtime-Support/issues/3013
Cannot say anything about iOS and 50.2.4.2-50.2.4.3 yet.

**Reply by FliplineStudios** [reply]
@premiumsB We're preparing a new iOS app with 50.2.4.3 (and switching to a Mac for packaging), mainly because of the fixes for audio issues on all iPhone 14/15 devices that were made with this version.  Also Apple is going to require iOS 17 / Xcode 15 for all builds in less than two months, so I think 50.2.4.1 would be the minimum needed for that.

**Reply by premiumsB** [reply]
@itlancer and @FliplineStudios Thank you for your reply. So basically you have to get a mac to build your ios app if not you may face major issue with iphone 14/15 or for the new requirement of apple with IOS 17.

Do @marchbold know when the new AIR version will be realeased with the patch for IOS app built with windows?

**Reply by ajwfrost** [reply]
The new AIR version with the patches (updated linker) to build iOS on Windows.. it's a complicated activity! It's in progress, but we've not managed to complete the first step yet = getting the Apple code to build on macOS. Once that's done, we'll have to try to then convert that into something that builds for Windows... and just a quick caveat: this would mean that the "stubs" mechanism that Adobe had been using will no longer work. Instead, the linking will need to be done against the official Apple iPhoneOS SDK, and we're not allowed to distribute that. So you would need to get hold of that somehow too. I know it's available online in various places but those are not official mechanisms... officially, you are only allowed to use the SDK on Apple-branded devices.

Basically, Apple want you to use their own platforms to develop software that's intended for their devices. We can try to make the process easier for those who want to use Windows as their development environment, but it may be that a mac machine will always then be needed to use the tools to interact with the Apple SDKs. We'd end up with a solution a little like Jenkins, where we have a "node" (client machine that's a mac) which is used by the server (the Windows PC) to carry out the build process.

**Reply by premiumsB** [reply]
ok @ajwfrost thank you for your answer.

**Reply by Michoko92** [reply]
thank you @ajwfrost . Right now, I'm using 50.2.3.4 on Windows, and all my builds seem to be accepted by Apple. So if I understood what @FliplineStudios wrote, we'll have to use at least 50.2.4.1 in two months to meet iOS 17 / Xcode 15 requirements but this version can't build IPAs properly on Windows? So to sum up, in 2 months, we'll either have to buy a Mac / build a Mac VM machine, and build on it? (except if you find how to build the new linker).  Or will 50.2.4.1 still work partially on Windows, given we don't use some features like fixed audio?

**Reply by premiumsB** [reply]
Hello @FliplineStudios 

I just bought an Iphone 14 and I tried 3 games apps that I have with the version 50.2.4.3 and I have 0 problem with anything. No bug sound no crash. I build my IPA with Animate 2023 and windows 11. 

Can you show us which bugs you have because I didn't see any bugs here.

Thank you

**Reply by FliplineStudios** [reply]
@premiumsB The sound issues should be fixed in 50.2.4.3, we've been testing the update on various iPhones and so far we haven't run into problems.  (Previously, sound would drop out and crackle on everything below 50.2.4.2, and there were crashes in 50.2.4.2 when playing multiple sounds which seems to be fixed now.)

The main issue with Windows builds is that Apple may reject the IPA when you upload it to App Store Connect (though I did see someone mention they were able to upload, so not sure if this issue comes and goes, or if it's based on ANEs you're using, etc.).  They'll install fine to your iPhone if you're installing locally, but Apple may not accept the Windows-packaged IPA for the App Store and will need a Mac-packaged one instead.

**Reply by premiumsB** [reply]
Oh ok I didn't upload a game yet but I will try and see if Apple accept it or not. Thanks for the follow-up.

**Reply by ajwfrost** [reply]
Trying to summarise it: there are currently two potential issues when using the AIR SDK to build an IPA on Windows:

1. the LLVM linker seems to result in a binary with different symbols defined which has caused Apple to reject some of these during the App Store review process. Some seem to be accepted, so I'm not sure what the difference are. The error message suggests we're using internal APIs, but we don't believe that to be the case. So, for this, it seems a bit hit and miss.

2. the way we link - against "stub" files which are dynamic libraries that we generate from the actual iPhoneOS SDK files - means that we don't get the nuances of which symbols may be in which libraries on which devices. Apple have moved symbols between different frameworks on different iOS versions. This means that there is a possible instability when the runtime tries to execute those symbols on an older iPhone. I don't believe any of these symbols are from the core AIR runtime, it seems like this is something we see more with ANEs that include third party frameworks.

The solution for all this would be to link using the official Apple linker, using the official Apple iPhoneOS SDK files (.tbd format).

The linker is, finally, open-sourced, so in theory we can build a version for Windows. We're struggling to even build it for macOS at the moment due to the dependencies that it comes with.. The iPhoneOS SDK we won't be able to distribute but there are ways that you could get this, albeit only via unofficial channels.

We are thinking that the better way to go - both from a timescale perspective, and from a general maintainability view - is to use a mac for this final linking stage. Some options would include being able to connect to a local mac on which you've got the necessary environment set up (like the Xamarin "Pair to Mac" feature), or potentially to be able to run this via a third party service online. We are exploring whether we could arrange/host such a service and make that available to the AIR SDK tools, to make it a little more seamless.

Hope that's useful info!

On a more general note, we're about to release 50.2.4.4 which should address a number of the recent issues with instabilities/build tool problems.

**Reply by Michoko92** [reply]
Thank you Andrew for the clarification. I am currently using a VMWare Sonoma MacOS VM for various tasks (for example to upload the IPA to the store). If you think connecting to a VM would do the trick, I suppose it could be an acceptable option (even if it would mean losing one of the most amazing features of this SDK, which is quite sad). Of course, offering an online service that would make everything transparent for Windows users would definitely be a very comfortable option for us. Good luck with this challenging issue! :)

**Reply by Michoko92** [reply]
@ajwfrost just FYI, I submitted yesterday an app to the Apple store, built with AIR 50.2.4.3 on Windows. The app was reviewed and accepted as normal. Here are the ANEs I use:

com.distriqt.Core
com.distriqt.InAppBilling
com.distriqt.Share
com.distriqt.Application
com.distriqt.ApplicationRater
com.distriqt.Dialog

So at least it seems AIR can still be used as is on Windows, for certain types of applications. Cheers!

**Reply by premiumsB** [reply]
Same thing here. I created an IPA with windows 11 and Adobe Air SDK 50.2.4.3 and apple accepted my IPA.
I use these ANEs:

com.distriqt.Core
com.distriqt.InAppBilling
com.distriqt.Share
com.distriqt.Application
com.distriqt.ApplicationRater
com.distriqt.Dialog
com.distriqt.Adverts
com.distriqt.playservices.Ads
com.distriqt.playservices.Base
com.distriqt.IDFA
com.distriqt.NetworkInfo
com.distriqt.AppGroupDefaults

@FliplineStudios Which ANEs do you use? How do you integrate your sound in your games?

**Reply by FliplineStudios** [reply]
@premiumsB We use distriqt ANEs as well.  We could try uploading a new Windows build to App Store Connect to see if it goes through, though we did already switch everything over to a Mac for packaging (planning to keep using this just in case we ran into any symbol/framework oddness on older iOS devices)

**Comment by premiumsB** [comment]
@FliplineStudios Which ANEs do you use exactly? Do you use this one com.distriqt.MediaPlayer.ane by any chance?

**Reply by FliplineStudios** [reply]
We use Core, InAppBilling, and SystemGestures.  For sound we just use the AS3 Sound/SoundChannel with all files internal in the library, which seems to be working great now with 50.2.4.3.

**Comment by Mintonist** [comment]
> On a more general note, we're about to release 50.2.4.4 which should address a number of the recent issues with instabilities/build tool problems.

@ajwfrost So if we can successfully build and deploy to AppStore our projects now - does it mean that problems have been solved for Windows or not?)

**Reply by ajwfrost** [reply]
I have to say, I'm still a little concerned about the Windows builds. If Apple have updated their checks so that they are no longer falsely flagging these LLVM-based binaries as containing private APIs, then that sounds great .. but we still likely have an issue with some iPhone APIs moving between components depending on the iOS version.

If you're able to extensively test your applications on older iOS versions, and it all works fine, then yes it's probably okay. But I believe there still some third party frameworks that are included in via ANEs that would then cause a crash on older iOS versions, because of the fact we link against generated stub .dylib files, rather than against the iPhoneOS SDK .tbd files.

So .. perhaps case by case; but for future-proofing, given how Apple have changed their linker mechanisms a few times now over the past couple of years, we are definitely looking at a feature where we can run the linker phase remotely i.e. the Windows machine pings the files over to a mac to do the link. Similar to Xamarin's "Pair to Mac" feature. We're also going to see whether we could provide a hosted service for this, although that may be something we look at later on if enough people need it.

thanks

**Reply by Mintonist** [reply]
> We're also going to see whether we could provide a hosted service for this, although that may be something we look at later on if enough people need it.

+1
